# API 資料串接 - 以 Dcard API 實作範例

* 了解 Dcard API 使用方式與回傳內容
* 撰寫程式存取 API 且解析 JSON 格式資料


## 作業目標

* 請利用 API: https://www.dcard.tw/_api/forums/pet/posts?popular=true 回答下列問題：

1. 這個 API 一次會回傳幾筆資料？每一筆資料包含哪些欄位？
2. 取出每一筆資料的「標題」、「貼文時間」、「留言人數」、「按讚人數」
3. 計算熱門/非熱門文章的「平均留言人數」與「平均按讚人數」

In [1]:
# 1. 這個 API 一次會回傳幾筆資料？每一筆資料包含哪些欄位？

# 個人備註
#這邊有個%s方法以後可以試試看
#==================================================

import requests
import json

r = requests.get('https://www.dcard.tw/_api/forums/pet/posts?popular=true')
js=json.loads(r.text)

count=0
for i in js:
    print(i['id'])
    count+=1
print('總共回傳%s資料'%count)




#答:一次共回傳30筆資料，

233149120
233151163
233148414
233154352
233153760
233153611
233154474
233153497
233153070
233155189
233155157
233152743
233148718
233149224
233148831
233154574
233153692
233155442
233149890
233148415
233155647
233155823
233154752
233151405
233149191
233148275
233154180
233155638
233151453
233156271
總共回傳30資料


In [2]:
# 個人備註
#因為json是字典形式(key value)，所以用 key value的形式去找出有幾個keys
#==================================================

print('每一筆資料包含以下欄位:')
num=js[1].keys()
count_1 = 0
for i in num:
    print(i)
    count_1+=1

print('共有%s個欄位'%count_1)

每一筆資料包含以下欄位:
id
title
excerpt
anonymousSchool
anonymousDepartment
pinned
forumId
replyId
createdAt
updatedAt
commentCount
likeCount
withNickname
tags
topics
meta
forumName
forumAlias
gender
school
replyTitle
mediaMeta
reactions
hidden
customStyle
isSuspiciousAccount
layout
withImages
withVideos
media
reportReasonText
postAvatar
共有32個欄位


In [3]:
# 個人備註
#
#==================================================

In [4]:
# 2. 取出每一筆資料的「標題」、「貼文時間」、「留言人數」、「按讚人數」

# 個人備註
#title
#createdAt
#commentCount
#likeCount
#==================================================
count = 1
for i in js:
    print(count,'標題: ',i['title'])
    count+=1
    print('貼文時間: ',i['createdAt'])
    print('留言人數: ',i['commentCount'])
    print('按讚人數: ',i['likeCount'])
    print('================')

1 標題:  #圖 在公園遛天竺鼠ヽ(✿ﾟ▽ﾟ)ノ
貼文時間:  2020-02-27T08:00:49.162Z
留言人數:  12
按讚人數:  188
2 標題:  拍屁屁奴隸的心聲
貼文時間:  2020-02-27T13:38:01.565Z
留言人數:  8
按讚人數:  96
3 標題:  習慣躺枕頭睡覺
貼文時間:  2020-02-27T05:50:41.877Z
留言人數:  9
按讚人數:  52
4 標題:  曬照🐿️
貼文時間:  2020-02-28T03:46:37.460Z
留言人數:  3
按讚人數:  47
5 標題:  文長 我擁有全世界最棒的貓咪
貼文時間:  2020-02-27T23:47:16.852Z
留言人數:  5
按讚人數:  40
6 標題:  來曬一下家裡的黑狗🐶
貼文時間:  2020-02-27T21:10:15.772Z
留言人數:  6
按讚人數:  32
7 標題:  三腳黑妞；救援孕狗不是件容易的事
貼文時間:  2020-02-28T04:16:05.721Z
留言人數:  1
按讚人數:  31
8 標題:  朕想怎麼睡 就怎麼睡(˶‾᷄ ⁻̫ ‾᷅˵)
貼文時間:  2020-02-27T19:53:48.404Z
留言人數:  0
按讚人數:  29
9 標題:  賤人！
貼文時間:  2020-02-27T17:57:19.700Z
留言人數:  3
按讚人數:  29
10 標題:  所以說到底是什麼生物呢
貼文時間:  2020-02-28T06:28:07.294Z
留言人數:  2
按讚人數:  25
11 標題:  求這篇限時的貓咪
貼文時間:  2020-02-28T06:23:06.364Z
留言人數:  0
按讚人數:  24
12 標題:  #協尋紅貴賓#懸賞金$5000#請大家幫忙頂上去！
貼文時間:  2020-02-27T17:05:28.180Z
留言人數:  3
按讚人數:  23
13 標題:  當你家的貓在看風景時⋯
貼文時間:  2020-02-27T06:48:59.523Z
留言人數:  0
按讚人數:  21
14 標題:  來曬一下可爾必思
貼文時間:  2020-02-27T08:19:49.020Z
留言人數:  1
按讚人數:  20
15

In [5]:
# 個人備註
#PANDAS方法
#=================================================

import pandas as pd

data = pd.DataFrame(js)

#選取我們需要什麼東西
allineed = data[['title','createdAt','commentCount','likeCount']]
#將我們需要的東西改個名字
allineed.columns = ['標題:','貼文時間:','留言人數:','按讚人數:']
#呈現出來
allineed

標題:                     貼文時間:  留言人數:  按讚人數:
0          #圖 在公園遛天竺鼠ヽ(✿ﾟ▽ﾟ)ノ  2020-02-27T08:00:49.162Z     12    188
1                    拍屁屁奴隸的心聲  2020-02-27T13:38:01.565Z      8     96
2                     習慣躺枕頭睡覺  2020-02-27T05:50:41.877Z      9     52
3                        曬照🐿️  2020-02-28T03:46:37.460Z      3     47
4              文長 我擁有全世界最棒的貓咪  2020-02-27T23:47:16.852Z      5     40
5                  來曬一下家裡的黑狗🐶  2020-02-27T21:10:15.772Z      6     32
6            三腳黑妞；救援孕狗不是件容易的事  2020-02-28T04:16:05.721Z      1     31
7      朕想怎麼睡 就怎麼睡(˶‾᷄ ⁻̫ ‾᷅˵)  2020-02-27T19:53:48.404Z      0     29
8                         賤人！  2020-02-27T17:57:19.700Z      3     29
9                 所以說到底是什麼生物呢  2020-02-28T06:28:07.294Z      2     25
10                   求這篇限時的貓咪  2020-02-28T06:23:06.364Z      0     24
11  #協尋紅貴賓#懸賞金$5000#請大家幫忙頂上去！  2020-02-27T17:05:28.180Z      3     23
12                當你家的貓在看風景時⋯  2020-02-27T06:48:59.523Z      0     21
13                   來曬一下可爾必思  2020-02-27T08:19:49.020Z      1     20
14                     露屁屁外星人  2020-02-27T07:09:01.168Z      0     15
15                     柯基犬曬萌照  2020-02-28T04:37:56.515Z      0     14
16                   看那顫抖的小眼眶  2020-02-27T22:52:19.976Z      3     14
17                同學我要妮的貓貓！！！  2020-02-28T07:26:54.480Z      1     13
18                       家有小三  2020-02-27T10:14:48.735Z      0     13
19                        🐔抓癢  2020-02-27T05:50:48.792Z      3     10
20                  懇請大家協尋狗小孩  2020-02-28T08:09:39.155Z      0      9
21                    #圖 酒醉大鼠  2020-02-28T08:49:27.019Z      2      8
22                 戲精貓-瞳孔放大片篇  2020-02-28T05:13:43.720Z      0      8
23               [抒發文]小狗狗去天堂了  2020-02-27T14:08:40.800Z      3      7
24             從小養到大的主子真的信任我嗎  2020-02-27T08:13:49.089Z      6      7
25                   學妹揪我一起看貓  2020-02-27T05:18:10.961Z      4      7
26                 尋狗 新莊夜市走丟的  2020-02-28T02:57:24.261Z      0      6
27                       虎視眈眈  2020-02-28T08:06:52.057Z      4      5
28                        彈力柴  2020-02-27T14:15:42.605Z      0      5
29               感覺是走失的狗狗🐕（？）  2020-02-28T10:27:25.209Z      2      4

In [6]:
# 3. 計算熱門/非熱門文章的「平均留言人數」與「平均按讚人數」
# 個人備註
#https://www.dcard.tw/_api/forums/pet/posts?popular=true  這是熱門
#https://www.dcard.tw/_api/forums/pet/posts?popular=false 這是非熱門即最新
#發現他API的網址的邏輯後就可以獲得不同內容了
#可看這篇https://levirve.github.io/blog/2016/Dccard-crawler/

#重要
#如果在改columns的時候直接用:
#latestdata.columns=['標題:','貼文時間:','留言人數:','按讚人數:']會有bug說:
#Pandas doesn't allow columns to be created via a new attribute name
#但是如果建立一個新變數rename2 = ['標題:','貼文時間:','留言人數:','按讚人數:']
#再latestdata.columns=rename2好像就沒事了
#https://stackoverflow.com/questions/52129876/userwarning-pandas-doesnt-allow-columns-to-be-created-via-a-new-attribute-name
#==================================================


latest = requests.get('https://www.dcard.tw/_api/forums/pet/posts?popular=false')
latestjson = json.loads(latest.text)

latestdata = pd.DataFrame(latestjson, columns=['title','createdAt','commentCount','likeCount'])

rename2 = ['標題:','貼文時間:','留言人數:','按讚人數:']

latestdata.columns=rename2
latestdata

標題:                     貼文時間:  留言人數:  按讚人數:
0              7公斤飼料大概長寬高?  2020-02-28T16:05:05.588Z      0      0
1   家有兩隻貓(⌯͒⁍̩̩᷄ ɪ ⁍̩̩᷄ฅ͒)  2020-02-28T16:04:47.921Z      0      2
2                   送養貓兄弟🐱  2020-02-28T16:00:39.989Z      0      1
3         #問 野生鼠會跟寵物鼠有共鳴嗎？  2020-02-28T15:39:29.606Z      0      0
4       雖然不是第一次，但還是很緊張的假新手  2020-02-28T15:27:26.396Z      0      0
5                  浪浪愛撒嬌❤️  2020-02-28T15:08:46.506Z      0      2
6               請問單兵該如何處置？  2020-02-28T14:59:10.360Z      2      2
7             轉：寵物美容店真的要慎選  2020-02-28T14:50:08.669Z      5      0
8              長輩說 兩隻狗 = 哭  2020-02-28T14:02:15.761Z      6      1
9                  真的好萌好萌😍  2020-02-28T12:51:56.196Z      0      2
10     #找主人  朝陽科大7-11附近走失貓  2020-02-28T12:43:11.635Z      0      2
11                 我家有一隻謎貓  2020-02-28T12:28:01.691Z      4      2
12                 #問 灰貓名字  2020-02-28T12:09:18.801Z      4      2
13            #貓 九州貓之島「相島」  2020-02-28T10:57:26.219Z      0      1
14            感覺是走失的狗狗🐕（？）  2020-02-28T10:27:25.209Z      2      4
15                   看過醫生了  2020-02-28T10:22:02.363Z      2      0
16               Costco狗飼料  2020-02-28T09:25:53.455Z      4      1
17                 #圖 酒醉大鼠  2020-02-28T08:49:27.019Z      2      8
18               懇請大家協尋狗小孩  2020-02-28T08:09:39.155Z      0      9
19                    虎視眈眈  2020-02-28T08:06:52.057Z      4      5
20                  你肯定有吸！  2020-02-28T07:40:49.986Z      0      4
21             同學我要妮的貓貓！！！  2020-02-28T07:26:54.480Z      1     13
22                 貓一出門就大叫  2020-02-28T07:19:23.699Z      1      4
23            被野貓咬要去打狂犬疫苗嗎  2020-02-28T07:00:46.555Z      2      2
24             所以說到底是什麼生物呢  2020-02-28T06:28:07.294Z      2     26
25                求這篇限時的貓咪  2020-02-28T06:23:06.364Z      0     24
26              戲精貓-瞳孔放大片篇  2020-02-28T05:13:43.720Z      0      8
27                  柯基犬曬萌照  2020-02-28T04:37:56.515Z      0     15
28        三腳黑妞；救援孕狗不是件容易的事  2020-02-28T04:16:05.721Z      1     31
29                    曬照🐿️  2020-02-28T03:46:37.460Z      3     47

In [7]:
# 3. 計算熱門/非熱門文章的「平均留言人數」與「平均按讚人數」

# 個人備註

#接下來熱門/非熱門都已經用好了
#.mean()的方法(這方法不知道是不是「已處理過的數據」的平均值算法)
#statistics.mean(data)
#https://docs.python.org/zh-tw/3/library/statistics.html#statistics.mean

#還有round的方法取四捨五入: round(x,y)
#==================================================
print('熱門文章的「平均留言人數」:',round(allineed[('留言人數:')].mean(),2))
print('熱門文章的「平均按讚人數」:',round(allineed[('按讚人數:')].mean(),2))
print('非熱門文章的「平均留言人數」:',round(latestdata[('留言人數:')].mean(),2))
print('非熱門文章的「平均按讚人數」:',round(latestdata[('按讚人數:')].mean(),2))


熱門文章的「平均留言人數」: 2.7
熱門文章的「平均按讚人數」: 26.73
非熱門文章的「平均留言人數」: 1.5
非熱門文章的「平均按讚人數」: 7.27


In [8]:
# 個人備註
#看到別人可以這樣用，也試試看
#'一段文字:{}'.format(想要運算的東西，結果丟到前面的{}裡面)
#==================================================
print('熱門文章的「平均留言人數」:{}'.format(round(allineed[('留言人數:')].mean(),2)))
print('熱門文章的「平均按讚人數」:{}'.format(round(allineed[('按讚人數:')].mean(),2)))
print('非熱門文章的「平均留言人數」:{}'.format(round(latestdata[('留言人數:')].mean(),2)))
print('非熱門文章的「平均按讚人數」:{}'.format(round(latestdata[('按讚人數:')].mean(),2)))

熱門文章的「平均留言人數」:2.7
熱門文章的「平均按讚人數」:26.73
非熱門文章的「平均留言人數」:1.5
非熱門文章的「平均按讚人數」:7.27


In [9]:
# 個人備註
#用numpy試試看，另一種取平均數的方式 
#np.mean([])
#==================================================
import numpy as np
print('熱門文章的「平均留言人數」:',round(np.mean(allineed[('留言人數:')]),2))
print('熱門文章的「平均留言人數」:',round(np.mean(allineed[('按讚人數:')]),2))
print('熱門文章的「平均留言人數」:',round(np.mean(latestdata[('留言人數:')]),2)) 
print('熱門文章的「平均留言人數」:',round(np.mean(latestdata[('按讚人數:')]),2)) 

熱門文章的「平均留言人數」: 2.7
熱門文章的「平均留言人數」: 26.73
熱門文章的「平均留言人數」: 1.5
熱門文章的「平均留言人數」: 7.27


# 第3題

答:

熱門文章的「平均留言人數」: 2.6

熱門文章的「平均按讚人數」: 24.2

非熱門文章的「平均留言人數」: 1.5

非熱門文章的「平均按讚人數」: 7.2